temporal-testing-junit4/src/main/java/io/temporal/testing/TestWorkflowRule.java

```diff
@@ -236,34 +242,29 @@ public void evaluate() throws Throwable {
     return watchman.apply(globalTimeout.apply(testWorkflowStatement, description), description);
   }
 
-  private void shutdown() {
-    testEnvironment.close();
-    if (interceptors != null) {
-      TracingWorkerInterceptor tracer = getInterceptor(TracingWorkerInterceptor.class);
-      if (tracer != null) {
-        tracer.assertExpected();
-      }
-    }
-  }
-
-  private void start() {
-    if (!doNotStart) {
-      testEnvironment.start();
-    }
-  }
-
   private String init(Description description) {
     String testMethod = description.getMethodName();
     String taskQueue = "WorkflowTest-" + testMethod + "-" + UUID.randomUUID().toString();
     Worker worker = testEnvironment.newWorker(taskQueue, workerOptions);
-    worker.registerWorkflowImplementationTypes(workflowTypes);
+    worker.registerWorkflowImplementationTypes(workflowImplementationOptions, workflowTypes);
     worker.registerActivitiesImplementations(activityImplementations);
     return taskQueue;
   }
 
-  /** Returns name of the task queue that test worker is polling. */
-  public String getTaskQueue() {
-    return taskQueue;
+  private void start() {
+    if (!doNotStart) {
```

Comment:
I see this flag has been added recently by Maxim, is it really needed or can we add ability to `addWorkflowImplementationFactory` in the rule?

Reply:
The idea was that some tests need to register test-specific workflows and activities which is not allowed after the start.

Reply:
I thought that we've agreed that having 1 to 1 mapping between rule and workflow types being tested is a good thing as it promotes clarity. From what I see the only thing that is missing in order to support what `DynamicWorkflowTest` needs is having ability to do `addWorkflowImplementationFactory` on the rule itself.